text/ProofModes.md

```diff
@@ -0,0 +1,118 @@
+# Proof Modes
+
+This CEP is about the infrastructure needed in order to have multiple 
+proof modes coexisting in a modular way (i.e. without syntactic or semantic conflicts)
+
+## Status quo (as per Coq 8.6)
+
+1. Tactic names are kernel names, i.e. unique, and can be accessed this way
+
+    ```
+    foo.tac arg1.
+    bar.tac arg1.
+    ```
+    
+1. Linking (not even requiring) a plugin loads into the *lexer* the tokens 
+   specific to the grammar implemented by the plugin
+   
+1. Linking (not even requiring) a plugin loads into the *parser* grammar the rules
+   specific to the plugin 
+   
+1. Macros in `tacextend.mlp` like `TACTIC EXTEND` plug the new grammar rule into the
+   *ltac* grammar `tactic` (non-terminal) symbol.  Note that tacticals are described by
+   the `tactic_expr` non-terminal, hence the use of `GEXTEND` in some plugins.
+   
+   - Today *ltac* is a plugin, and `TACTIC EXTEND` is tied to this particular plugin.
+     Tactic arguments as in `ARGUMENT EXTEND` are also part of this plugin.
+   
+   - `decl_mode`, for example, cannot use the `* EXTEND` mechanism because it needs to
+     plug in its own non-terminal (to hide ltac).
+
+1. So called "proof mode" is implemented by making the main "tactic" non-terminal
+   mutable and having specific commands assign such non-terminal to the desired one.
+   The current proof mode is not part of the `Summary` and is both stored in `Proof_global`
+   and `Stm` that do their best to ensure parsing happens in a state where such non-terminal
+   holds the expected value.
+
```

Comment:
The only trade off between Summary and STM is the following one.

In Summary: to set the right parsing state, one has to install a full system state (no work on the STM side, since we do/should-do this already).

In STM: more work, but one could set the right parsing state without even having computed the corresponding system state.  One could only process "notation commands", i.e. be faster.